cli-tools/sagebig/src/main.rs

```diff
@@ -97,3 +115,44 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
 
     Ok(())
 }
+
+fn cmd_extract(args: &clap::ArgMatches) -> Result<(), Error> {
+    let big_path = args.value_of("archive_path").unwrap();
+    let archive = try!(BigArchive::new_from_path(&big_path));
+    let output_dir = Path::new(args.value_of("output_dir").unwrap());
+    let verbose = args.is_present("verbose");
+
+    // Create output dir if it doesn't exist
+    if !output_dir.exists() {
+        try!(fs::create_dir(&output_dir).map_err(ExtractError::StdIoError));
+
+        if verbose {
+            println!("Created output_dir `{:?}`", &output_dir.display());
+        }
+    }
+
+    let output_dir = try!(output_dir.canonicalize().map_err(ExtractError::StdIoError));
+
+    for name in archive.get_all_entry_names() {
+        let entry = archive.get_entry(name)
+            .expect(&format!("Failed to read known entry {} from {}", name, big_path));
+
+        let file_path = output_dir.join(Path::new(&entry.name));
+        let file_dirpath = match file_path.parent() {
+            Some(x) => x,
+            None => return Err(Error::ExtractError(ExtractError::InvalidFilePath(file_path .to_string_lossy().into_owned()))),
```

Comment:
I think what you want is [`Path::is_file`}(https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.is_file) paired with a possible `Path::parent`.

I say possible because `foo.txt` in the root is completely valid.